src/cryptonote_core/uptime_proof.cpp

```diff
@@ -0,0 +1,111 @@
+#include "uptime_proof.h"
+#include "common/string_util.h"
+#include "version.h"
+
+extern "C"
+{
+#include <sodium/crypto_sign.h>
+}
+
+#undef OXEN_DEFAULT_LOG_CATEGORY
+#define OXEN_DEFAULT_LOG_CATEGORY "uptime_proof"
+
+//Default constructor for the uptime proof, will take the service node keys as a param and sign 
+uptime_proof::Proof::Proof(uint32_t public_ip, uint16_t storage_port, uint16_t storage_lmq_port, const std::array<uint16_t, 3> ss_version, uint16_t quorumnet_port, const std::array<uint16_t, 3> lokinet_version, const service_nodes::service_node_keys& keys)
+{
+  snode_version = OXEN_VERSION;
+  pubkey = keys.pub;
+  timestamp = time(nullptr);
+  public_ip = public_ip;
+  storage_port = storage_port;
+  pubkey_ed25519 = keys.pub_ed25519;
+  qnet_port = quorumnet_port;
+  storage_lmq_port = storage_lmq_port;
+  storage_version = ss_version;
+  this->lokinet_version = lokinet_version;
+  crypto::hash hash = uptime_proof::hash_uptime_proof(*this);
+
+  crypto::generate_signature(hash, keys.pub, keys.key, sig);
+  crypto_sign_detached(sig_ed25519.data, NULL, reinterpret_cast<unsigned char *>(hash.data), sizeof(hash.data), keys.key_ed25519.data);
+}
+
+//Deserialize from a btencoded string into our Proof instance
+uptime_proof::Proof::Proof(const std::string& serialized_proof)
+{
+  try {
+    lokimq::bt_dict bt_proof = lokimq::bt_deserialize<lokimq::bt_dict>(serialized_proof);
```

Comment:
You should `const` this, and then change all the `bt_proof["key"]` to `bt_proof.at("key")`.  The `[]` operator will auto-create fields that don't exist, while `.at()` will throw if the key doesn't exist, and since nearly all these fields are mandatory that seems desirable if a key is missing.